### PR TITLE
Feature/2-view-cacher

### PR DESCRIPTION
**Pull request by ayushbaid**

### CODE REVIEW

**Review by akshay-krishnan — COMMENTED** [review]
could we also use this class?

**Comment by ayushbaid** [comment]
> could we also use this class?

I dont want to use it on CI because we might run out of limited cached space.

**Comment by akshay-krishnan** [comment]
> > could we also use this class?
> 
> I dont want to use it on CI because we might run out of limited cached space.

do we have any non-CI configs? 
I think the size of that `results` variable will be much smaller than the correspondences for the same edge, which we cache. So it is definitely not the bottleneck for memory..

**Comment by akshay-krishnan** [comment]
@ayushbaid ping on this

**Comment by ayushbaid** [comment]
> @ayushbaid ping on this

Done!

**Comment by johnwlambert** [comment]
Hi @ayushbaid, thanks for adding this.

I think CI is failing because of these two errors:
```
=========================== short test summary info ============================
FAILED tests/test_gtsfm_runner_olssonloader.py::test_gtsfm_runner_olssonloader[dataset_path0] - AttributeError: 'TwoViewEstimatorCacher' object has no attribute '_verifier'
FAILED tests/test_gtsfm_runner_olssonloader.py::test_gtsfm_runner_olssonloader[dataset_path1] - AttributeError: 'TwoViewEstimatorCacher' object has no attribute '_verifier'
====== 2 failed, 425 passed, 7 skipped, 21 warnings in 790.29s (0:13:10) =======
Restarting worker
```

File: gtsfm/two_view_estimator_cacher.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+logger = logger_utils.get_logger()
+
+mpl_logger = logging.getLogger("matplotlib")
+mpl_logger.setLevel(logging.WARNING)
+
+pil_logger = logging.getLogger("PIL")
+pil_logger.setLevel(logging.INFO)
+
+
+class TwoViewEstimatorCacher(TwoViewEstimator):
```
Comment (johnwlambert) [inline gtsfm/two_view_estimator_cacher.py:32]: could we add a unit test for this?

File: gtsfm/two_view_estimator_cacher.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+"""Decorator which implements a cache for the two-view-estimator class
+
+Authors: Ayush Baid
+"""
```
Comment (johnwlambert) [inline gtsfm/two_view_estimator_cacher.py:4]: should this file go under gtsfm/frontend/cacher/two_view_estimator_cacher? @ayushbaid

File: gtsfm/two_view_estimator_cacher.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/two_view_estimator_cacher.py:36]: Added one

File: tests/test_two_view_estimator_cacher.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    @patch("gtsfm.utils.cache.generate_hash_for_numpy_array", return_value="numpy_key")
+    @patch("gtsfm.utils.io.read_from_bz2_file", return_value=None)
+    @patch("gtsfm.utils.io.write_to_bz2_file")
+    def test_cache_miss(
+        self, write_mock: MagicMock, read_mock: MagicMock, generate_hash_for_numpy_array_mock: MagicMock
+    ) -> None:
+        """Test the scenario of cache miss for the TwoViewEstimator."""
+
+        # Mock the underlying two-view estimator which is used on cache miss.
+        underlying_estimator_mock = MagicMock()
```
Comment (akshay-krishnan) [inline tests/test_two_view_estimator_cacher.py:46]: I think the `underlying` word is unnecessary..

File: tests/test_two_view_estimator_cacher.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        # Assert that read function was called once and write function was called once.
+        cache_path = ROOT_PATH / "cache" / "two_view_estimator" / "numpy_key.pbz2"
+        read_mock.assert_called_once_with(cache_path)
+        write_mock.assert_called_once()
+
+
+    @patch("gtsfm.utils.cache.generate_hash_for_numpy_array", return_value="numpy_key")
+    @patch("gtsfm.utils.io.read_from_bz2_file", return_value=_DUMMY_OUTPUT)
+    @patch("gtsfm.utils.io.write_to_bz2_file")
+    def test_cache_hit(
```
Comment (akshay-krishnan) [inline tests/test_two_view_estimator_cacher.py:80]: is this guaranteed to run after the previous test? I guess youre assuming that?

**Review by akshay-krishnan — APPROVED** [review]

File: tests/test_two_view_estimator_cacher.py (hunk 4 above)
Comment (johnwlambert) [reply tests/test_two_view_estimator_cacher.py:80]: The cache hit is actually ensure by the mocked return value of the `read_from_bz2_file()`. That functions returns None on cache miss. So the order of execution of functions doesnt matter here.

Comment (johnwlambert) [reply tests/test_two_view_estimator_cacher.py:80]: Added a comment to clarify this now.

File: tests/test_two_view_estimator_cacher.py (hunk 3 above)
Comment (johnwlambert) [reply tests/test_two_view_estimator_cacher.py:46]: Acknowledged, I think it's helpful : - )

File: gtsfm/two_view_estimator_cacher.py (hunk 2 above)
Comment (ayushbaid) [reply gtsfm/two_view_estimator_cacher.py:4]: Yes good idea.

**Pull request merged**